x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -237,4 +392,293 @@ public void onFailure(Exception e) {
+                        knownSnapshots = listSearchableSnapshots(state);
+                        existingSnapshots = knownSnapshots;
+                    }
+
+                    final Set<String> knownRepositories = state.metadata()
+                        .custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY)
+                        .repositories()
+                        .stream()
+                        .map(RepositoryMetadata::name)
+                        .collect(Collectors.toSet());
+
+                    Object[] lastSortValues = null;
+                    for (SearchHit searchHit : searchHits) {
+                        lastSortValues = searchHit.getSortValues();
+                        assert searchHit.getId() != null;
+                        assert searchHit.hasSource();
+
+                        try {
+                            final CachedBlob cachedBlob = CachedBlob.fromSource(searchHit.getSourceAsMap());
+                            if (cachedBlob.creationTime().isAfter(expirationTime)) {
+                                logger.trace("blob store cache entry with id [{}] was created recently, skipping", searchHit.getId());
+                                continue;
+                            }
+                            boolean delete = false;
+
+                            // See {@link BlobStoreCacheService#generateId}
+                            // doc id = {repository name}/{snapshot id}/{snapshot index id}/{shard id}/{file name}/@{file offset}
+                            final String[] parts = Objects.requireNonNull(searchHit.getId()).split("/");
+                            assert parts.length == 6 : Arrays.toString(parts) + " vs " + searchHit.getId();
+
+                            final String repositoryName = parts[0];
+                            if (knownRepositories.contains(repositoryName) == false) {
+                                logger.trace("deleting blob store cache entry with id [{}]: repository does not exist", searchHit.getId());
+                                delete = true;
+                            } else {
+                                final Set<String> knownIndexIds = knownSnapshots.get(parts[1]);
+                                if (knownIndexIds == null || knownIndexIds.contains(parts[2]) == false) {
+                                    logger.trace("deleting blob store cache entry with id [{}]: not used", searchHit.getId());
+                                    delete = true;
+                                }
+                            }
+                            if (delete) {
+                                final DeleteRequest deleteRequest = new DeleteRequest().index(searchHit.getIndex());
+                                deleteRequest.id(searchHit.getId());
+                                deleteRequest.setIfSeqNo(searchHit.getSeqNo());
+                                deleteRequest.setIfPrimaryTerm(searchHit.getPrimaryTerm());
```

Comment:
I wonder if this is necessary. Not that it harms, but we do not retry on failure and I think it would be correct without it. Obviously it will be sorted on next round in case it should fail.

Reply:
I added this to avoid collision in case of a primary shard failing over. But since we only delete and don't retry or log, it does not make much sense. I removed this in 1d0654f0d12f8432c045f668561630052f6435fc